p2panda-sync-next/src/traits.rs

```diff
@@ -0,0 +1,60 @@
+use std::fmt::Debug;
+use std::hash::Hash as StdHash;
+
+use futures::Sink;
+use futures_util::{AsyncRead, AsyncWrite, Stream};
+use serde::{Deserialize, Serialize};
+
+
+/// Sync protocol which runs over an AsyncWrite and AsyncRead pair.
+pub trait SyncProtocol {
+    type Output;
+    type Error;
+    type Event;
+
+    fn run(
+        self,
+        tx: &mut (impl AsyncWrite + Unpin),
+        rx: &mut (impl AsyncRead + Unpin),
+    ) -> impl Future<Output = Result<Self::Output, Self::Error>>;
+}
+
+// NOTE(sam): we don't strictly need this trait as it isn't used in the public APIs, but it's nice to
+// encourage uniformity across general re-usable protocol implementations. We can decide if we
+// like it or would rather remove it.
+/// Generic protocol interface which runs over a typed sink and stream pair. 
+pub trait Protocol {
+    type Output;
+    type Error;
+    type Message;
+
+    fn run(
+        self,
+        sink: &mut (impl Sink<Self::Message, Error = Self::Error> + Unpin),
+        stream: &mut (impl Stream<Item = Result<Self::Message, Self::Error>> + Unpin),
+    ) -> impl Future<Output = Result<Self::Output, Self::Error>>;
+}
+
+/// Identify the particular dataset a peer is interested in syncing.
+///
+/// Exactly how this is expressed is left up to the user to decide. During sync the "initiator"
+/// sends their topic query to a remote peer where it is be mapped to their local dataset.
+/// Additional access-control checks can be performed. Once this "handshake" is complete both
```

Comment:
Here it mentions "access control" checks? Are we doing this already? Is this part of the next PR?